lib/theme_check/language_server/completion_engine.rb

```diff
@@ -3,35 +3,23 @@
 module ThemeCheck
   module LanguageServer
     class CompletionEngine
-      include PositionHelper
-
-      def initialize(storage)
+      def initialize(storage, bridge = nil)
         @storage = storage
+        @bridge = bridge
         @providers = CompletionProvider.all.map { |x| x.new(storage) }
       end
 
       def completions(relative_path, line, col)
-        buffer = @storage.read(relative_path)
-        cursor = from_row_column_to_index(buffer, line, col)
-        token = find_token(buffer, cursor)
-        return [] if token.nil?
-
-        @providers.flat_map do |p|
-          p.completions(
-            token.content,
-            cursor - token.start
-          )
+        @providers.flat_map do |provider|
+          provider.completions(relative_path, line, col)
```

Comment:
Thank you, @charlespwd! Excellent point. I've introduced the `CompletionContext` object to avoid IO calls/token-finding.

Thus, now completion providers implement `completions(context)` instead of `completions(relative_path, line, col)` to support that :)